### ISSUE TITLE
Save/load files to and from cloud platforms

### ISSUE THREAD

**Issue by hschwentner** [WPS/egon.io#84]
**Is your feature request related to a problem? Please describe.**
Currently the Modeler allows files only to be saved and loaded from the device the browser is running on.

**Describe the solution you'd like**
Alternatively to the device I would like to save and load files to Google Drive, Dropbox, OneDrive, GitHub, and the like.

***Additional context**
Draw.io has a feature that is very similar. Maybe their solution could be used.

**Comment by hofstef** [comment]
I see some similarity to #68 . I guess the difference would that with #68, the files are shared with anyone who has the link whereas with #84 the file is uploaded to a personal cloud storage (requires authentication).

**Comment by almereyda** [comment]
Common cloud storage providers for this use case are:

- Google Drive
- Nextcloud (WebDAV)
- Apple iCloud
- Box

It would even be imaginable to integrate the App into Nextcloud similar to their Draw.io integration.

Less common, but certainly possible storage providers could be:

- [Remote Storage](https://remotestorage.io/)
- [Solid](https://solidproject.org/)

Involving OIDC, a token service and a lambda/daemon to generate pre-signed URLs, it would also be possible to integrate with S3.

**Comment by hofstef** [comment]
draw.io and others have built their integrations on the backend - but Egon is frontend only, so we have to take a different approach. Dropbox offers JavaScript-based embeddable components that seem to provide the required functionality: https://www.dropbox.com/developers/documentation?_tk=pilot_lp&_ad=altcta1&_camp=docs

We'll give it a try and do some research if other provides offer similar solutions.

**Comment by hofstef** [comment]
Google Drive: https://developers.google.com/drive/picker/guides/overview

**Comment by hofstef** [comment]
**TL;DR:** This feature would provide limited convenience for some cloud storage users. The effort for building and maintaining this, on the other hand, is rather high. So we are **not** going to build this.

### Let's break down possible use cases for this feature request:
1. I want my domain stories to be backed up, versioned and available on multiple devices. Storing them in the cloud provides that for me. An integration saves me the hassle of downloading a domain story to my hard drive and uploading it to my cloud provider.
2. If I have a domain story stored in the cloud, I want to import it directly into Egon so that I don't need to download and then import it from the local file system.
3. I want to share domain stories with a group of other Egon users so that everyone in our group can edit the domain stories (two-way sharing). 
4. I want to provide domain stories to other Egon users e.g., as a trainer I want to share examples with participants (one-way sharing).

### Now let's look at some constraints:
- **with sync:** Many cloud platforms enable synchronization of a local file system with the cloud (Dropbox, OneDrive, Google Drive, iCloud, ...).
- **without sync:** Some users do not synchronize their local file system with the cloud, either because they choose not to or because the cloud platform does not provide such a capability (e.g. GitHub).
- **full/restricted access:** Dropbox let developers choose if their app should have access to all folders (and allow to browse all folders, create folders,...) or to just one, app-specific folder. We assume that other cloud providers have similar mechanisms. Egon users might hesitate to give Egon full access to their cloud storage; we could settle for restricted access (which btw is what diagrams.net does) but I find it unlikely that someone wants to keep all their domain stories in just one folder.

### What we would need to build (an maintain):
- for every cloud platform, we need to register an API key and (possibly) get approval from the platform provider
- build services for every cloud platform we want to support
- provide sign-in and sign-out capabilities
- build a dialog that lets users choose folders and files for import and export
- integrate the feature in a way so that the user interface makes sense
- communicate clearly to users, what they agree to when they use the feature and what the limitations are
- security becomes a much bigger issue if Egon accesses the users' cloud storage
- watch for changes of cloud platform APIs
- add configuration options to Egon so that companies that host Egon themselves can turn cloud integrations off if they have security concerns

### Evaluation
The feature provides **little to no value** for use cases (1) and (2) when the Egon user uses a cloud platform **with sync**:
- Without the feature: You download your domain story to the browser's download folder and move it to the synced folder. To import, you need to navigate to your sycned folder.
- With the feature, you would still have to navigate to the folder you want the file to be saved to/loaded from.

If the user does not sync their file system to a cloud provider, the feature has **little value** for use case (1) as it would only spare the user to upload a file to their cloud provider. For use case (2) it would provide **little to moderate value**, but still, it's only convenience. If the domain story is in a publicly shared folder, it might be easier to use the "import from URL" feature (#68).

Use case (3) extends use cases (1) and (2). It would require the solution to have access to the full online folder structure to be usable (I don’t think sharing a single app-specific folder with all your domain stories in it is a good idea). How much value the feature provides is again dependent on the usage of folder synchronization.

Use case (4) also extends use cases (1) and (2).  If the domain stories are in a publicly shared folder, the "import from URL" feature (#68) seems like a good alternative.
